### ISSUE TITLE
When a community is selected, the symbology should immediately switch to classification based on community class breaks

### ISSUE THREAD

**Issue by torrin47** [Innovate-Inc/EnviroAtlas#357]
Allie and Jessica will be generating JSON files containing the pre-calculated rendering information so we won't need to calculate it on the fly. We will need to accept those JSON files and set the rendering accordingly.

**Comment by torrin47** [comment]
From: Mackey, Alexandra 
Sent: Wednesday, March 29, 2017 12:22 PM
To: Moore, Rose-Marie <Moore.Rose-Marie@epa.gov>; Neale, Anne <Neale.Anne@epa.gov>; Pierson, Suzanne <Pierson.Suzanne@epa.gov>; Rosenbaum, Barbara <Rosenbaum.Barbara@epa.gov>; Hultgren, Torrin <Hultgren.Torrin@epa.gov>; Ji, Baohong <Ji.Baohong@epa.gov>; Yarnell, David <Yarnell.David@epa.gov>; Daniel, Jessica <Daniel.Jessica@epa.gov>; Baynes, Jeremy <baynes.jeremy@epa.gov>
Subject: Community Symbology JSON

Hey all, 

I’ve finally finished up writing all the JSON files for the community legends. They are located on the FTP in the folder: /EnviroAtlas/ForAli/CommSelection_JSON_20170324.zip. I included the table with the -888888 values added here: CommSelection_JSON_20170324.zip\Full.gdb\CombComm_BG_Alb_8s

I wanted to take this time to review all the pieces of the community selection widget to make sure we’re all on the same page with the functionality before we start implementing the JSON, etc. You’ll notice each item has 1, 2, or 3 stars after it. These indicate whether the item is an absolute need(***), something we really want to see but could live without if we have to(**), or something that we would like to see but aren’t married to(*). I’ve tried to order the list in same order as the user will interact with the widget and the data. 

1.	When a community data layer is selected and turned on and a community has been selected: the legend reflects the breaks specific to that community. ***
2.	Whenever a community data layer is turned on ALL the data for all the communities are visible and conform to the legend of the selected community. **
3.	When a community data layer has NULL values, the labels for those values must accurately reflect why the BG is a NULL. ***
4.	Community data layers must be able to display multiple types of nulls in the legend. It is unlikely that we will need more than 2 null labels in a single layer in the near future, but I can’t make any promises. I’ve added some examples below. ***
5.	Because we will be displaying all the communities, non-selected communities may have data outside of the range of the selected community. We would like the legend to say “Out of Range” in this case but this is not reflected in the JSON. I can add it if needed and someone can tell me how it should fit in the format.*
6.	The only time the legend changes is when triggered by the community selection widget (and/or the points layer you’re working on). Not by panning, bookmarks, or searching a location. ***
7.	The attribute table should only have the values for the selected community. *
8.	The change symbology widget should work within the data for the selected community. We will need to discuss how the change symbology will work with null values. Our preference would be that the null values retain their EnviroAtlas-defined symbology. ***


A few notes about the JSON:
•	Wet_M is missing for a few communities. This isn’t really an issue since these data aren’t displayed in the map, just something to note. 
•	I’m going to run WVW_Pct and WVW_Pop for PAZ and DNC so those won’t be empty. I also think I have data somewhere for POR day cares that I might be able to use. 
•	The JSON includes all the communities through MSPMN but does not include NHCT, so there will be an updated version before we go live; it will include the above changes along with NHCT. 
•	For the JSON itself: I’ve given you 4 options to work with to figure out what is the most efficient. 
1.	Layers Individually, WITH NULLS: [Comm]_JSON_Symbol/Nulls/[Comm]_[Layer].json  Torrin wanted all the layers for all the communities in separate JSON files + these JSON files incorporate the NULL values (listed below) and their appropriate labels. 
2.	Layers combined, WITH NULLS: [Comm]_JSON_Symbol/Nulls/[Comm]_Full.json  Barb and I wanted to test using one JSON file per communtiy + these JSON files incorporate the NULL values (listed below) and their appropriate labels.
3.	Layers Individually, WITHOUT NULLS: [Comm]_JSON_Symbol/NoNulls/[Comm]_[Layer].json  Torrin wanted all the layers for all the communities in separate JSON files + these JSON files essentially ignore the NULL values and labels; those would have to be defined elsewhere in the programming. 
4.	Layers combined, WITHOUT NULLS: [Comm]_JSON_Symbol/NoNulls/[Comm]_Full.json Torrin wanted all the layers for all the communities in separate JSON files + these JSON files essentially ignore the NULL values and labels; those would have to be defined elsewhere in the programming. 
•	As stated before, the nulls MUST be labelled in the legend. I’m sending along JSON without the nulls because, in the flex version, the nulls are defined/labelled through your coding outside of the symbology. If it is easier if they are incorporated into the JSON, please let me know.  Whichever route we take, we must be able to have 2+ null labels. 
•	I don’t care which option you use as long as the functions in the list above still work. Once you figure out which works best, I will only produce that version. 
•	If there is anything that needs to change in the JSON or in the naming conventions (or anything else), let me know and I will update it before doing the next run. 
•	I recommend that when I send new data, I will resend ALL the JSON and you can just replace the whole folder. That way you won’t have to dig through and replace individual files – especially important if we end up going with the individual layers option.
•	If a community doesn’t have a data layer, the legend is set to match the combined communities legend, and will need to be replaced every time the data changes. 

Lastly, a few notes about NULLs, so we’re all on the same page:

•	Nulls are very important when a zero value has a very different meaning than a Null/NoData value: For example, having 0% vegetation in the riparian buffer is very different than having no riparian buffer area for which to calculate percent vegetation. 
•	-888888: “No Data for Community”  Occurs when a particular layer is not calculated for an entire community (Ag_P)
•	-99999:
o	Primary Nulls: These occur frequently, by individual BG; sometimes in the same layer as -888888. If we change these to separate values, we would need updated all the data AND METADATA! 
o	“No Day Cares”  Day_Low
o	“No Schools”  K12_Low
o	“No Buffer Area”  RB15_ForP, RB15_ImpP, RB15_VegP, RB50_ForP, RB50_ImpP, RB50_VegP
o	“No People Living in Block Group”  MFor_PC, Buff_Pop, Buff_Pct, SBuff_Pop, IBuff_Pop, Green_PC, Imp_PC, IWDP_Pop, IWDP_Pct, BWDP_Pop, BWDP_Pct, WVT_Pop, WVT_Pct, WVW_Pop, WVW_Pct, Ag_PC [soon to include a floodplain layer]
o	“No Floodplain Area”  [coming soon]
o	If it would help, we can change these such that each label has its own value, but this would involve changing the metadata for all the layers in all the communities and probably changing the download data too. 
•	-9999X:
o	Secondary Nulls: These occur less frequently and often in combination with -99999 and -888888; These don’t actually exist in our data yet, but will be coming with Chicago and the Floodplain layers
o	“No Data for Block Group”  [coming soon]; Will occur because the LC from Jarlath only goes to the edge of the community so some of our metrics won’t be calculated for BGs around the edges of the communities that Jarlath completes. 
o	“No FEMA Floodplain Data”  [coming soon]; 

I think that is everything related to community selection. Sorry for the super long email! 

Thanks,
Ali 

Alexandra Sears Mackey
EPA ORISE Participant
GIS Analysis | EnviroAtlas
Office of Research and Development
919-541-4350
mackey.alexandra@epa.gov

**Comment by Baohong** [comment]
I did some research on rendering community layers. There is one called UniqueValueRenderer which could combine two fields/attributes for rendering. In our EnviroAtlas case, the two fields are "CommST" and "Ag_M" (or other designated field). I hoped that our pre-defined symbology could appear on map for all communities at the same time, so that we do not have to remove / re-add community layers when user choose another community. But, since UniqueValueRenderer only deal with a single unique value for each field rather than a range, I did not find any code to incorporate class breaks into UniqueValueRenderer. Currently I cannot use such method. 
So I might still have to use classic classbreaks for rendering. This involves removing a community layer when user select a different community and adding it back. Then there will a problem of the current attribute table tab disappearing from table tabs. User have to re-open attribute table when he selects another community.
